[sc-146432] Refactor build CI to properly verify plugin compatibility

jazanne opened the pull request:
Update GitHub actions to verify plugin compatibility

I had to refactor the build & prerelease jobs because the verify action would fail with the up/downloaded artifact (it was looking in the wrong place, a limitation of the github action). Also, it looks like the **Verify Plugin** job has been [failing silently](https://github.com/launchdarkly/ld-intellij/runs/5310288820?check_suite_focus=true) for a long time

This is what a compatibility failure looks like:
[See workflow run](https://github.com/launchdarkly/ld-intellij/runs/5601184509?check_suite_focus=true)

You have to expand the logs above this message to see the exact errors

```
Error: ==============================================
Error: ==============================================
Error: ===                                        ===
Error: ===    PLUGIN FAILED VERIFICATION CHECK    ===
Error: ===                                        ===
Error: ==============================================
Error: ==============================================
Error: Exiting due to a known, handled, plugin validation failure.
```

Example error output
```
 Plugin com.github.intheclouddan.intellijpluginld:0.5.1-idea22 against IC-213.5744.223: 2 compatibility problems. 72 usages of experimental API
  Compatibility problems (2): 
      #Invocation of unresolved method com.intellij.ui.dsl.builder.Row.DefaultImpls.text$default(Row, String, int, HyperlinkEventAction, int, Object) : Cell
          Method com.launchdarkly.intellij.settings.LaunchDarklyApplicationConfigurable.createPanel$1$1.invoke(com.intellij.ui.dsl.builder.Row $this$row) : void contains an *invokestatic* instruction referencing an unresolved method com.intellij.ui.dsl.builder.Row.DefaultImpls.text$default(com.intellij.ui.dsl.builder.Row, java.lang.String, int, com.intellij.ui.dsl.builder.HyperlinkEventAction, int, java.lang.Object) : com.intellij.ui.dsl.builder.Cell. This can lead to **NoSuchMethodError** exception at runtime.
      #Invocation of unresolved method com.intellij.ui.dsl.builder.Row.DefaultImpls.comment$default(Row, String, int, HyperlinkEventAction, int, Object) : Cell
          Method com.launchdarkly.intellij.settings.LaunchDarklyApplicationConfigurable.createPanel$1$3.invoke(com.intellij.ui.dsl.builder.Row $this$row) : void contains an *invokestatic* instruction referencing an unresolved method com.intellij.ui.dsl.builder.Row.DefaultImpls.comment$default(com.intellij.ui.dsl.builder.Row, java.lang.String, int, com.intellij.ui.dsl.builder.HyperlinkEventAction, int, java.lang.Object) : com.intellij.ui.dsl.builder.Cell. This can lead to **NoSuchMethodError** exception at runtime.
```

Refactor **does**
- consolidates jobs and removes duplicated steps, so it should be quicker
- updates versions of github actions used
- removes compatibility for 2022/EAP because of UI DSL issues

Refactors **doesn't**
- reduce duplication between build & prerelease steps - there are some open comments / notes about refactoring the release process so that it's more of a one-click experience, so I held off on any further changes. My primary objective was getting the verify plugin step to work.


Ensure we verifying across correct compatibility range:
https://github.com/launchdarkly/ld-intellij/blob/4820773ec1862c3257147b37c454d1e2b6587c52/gradle.properties#L3-L13

jazanne committed:
update ide versions being verified

carmenquan commented on .github/workflows/ide_versions_file.txt:
```diff
@@ -0,0 +1,4 @@
+ideaIU:2021.3
```
smallest nit: can we do
```
ideaIC:2021.3
ideaIU:2021.3
ideaIC:2022.1
ideaIU:2022.1
```

jazanne replied:
lol yea that's my bad!

jazanne committed:
update verify action and options

jazanne committed:
need to use checkout

jazanne committed:
latest eap

jazanne committed:
update artifcat up/download

remove release draft

jazanne committed:
encouraged to use github token

jazanne committed:
update action

setup-java now requires distribution
zulu is recommended for easy migration from v1

jazanne committed:
update plugin location

jazanne committed:
update plugin location again

jazanne committed:
checkout v3

jazanne committed:
print build outputs

jazanne committed:
test line

jazanne committed:
fix typo

jazanne committed:
lint

jazanne committed:
remove test step

jazanne committed:
another thing

jazanne committed:
trying again...

jazanne committed:
try merging test, build, verify...

to avoid upload/download

jazanne committed:
revert prerelease for now

jazanne committed:
make sure logs print

jazanne committed:
skip some steps for testing

jazanne committed:
once again remove optional props

jazanne committed:
output logs is bad

jazanne committed:
add pre release to pr for testing

jazanne committed:
re enable build steps

jazanne committed:
update cache

jazanne committed:
update prerelease

jazanne committed:
Revert "add pre release to pr for testing"

This reverts commit c250f3628d2a47c87ed6f2cea6565c0470fda408.

jazanne committed:
cleanup

jazanne committed:
Merge remote-tracking branch 'origin/main' into jwhite/sc-146432/run-plugin-verifier-in-ci-to-check-for-compatibility

jazanne commented on .github/workflows/build.yml:
```diff
@@ -120,96 +94,22 @@ jobs:
-      - name: Remove Old Release Drafts
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: Summary
```
added this because it's nice to debugging

jazanne committed:
try resolving compatibility issues

jazanne committed:
use cell comment

jazanne committed:
revert layout changes

jazanne committed:
only support 2021 IDEs

jazanne commented on gradle.properties:
```diff
@@ -4,7 +4,7 @@ pluginGroup=com.github.intheclouddan.intellijpluginld
 pluginVersion=0.5.1-idea22
 pluginSinceBuild=213.6461.79
-pluginUntilBuild=221.*
+pluginUntilBuild=213.*
```
our compatibility issues stem from 221 (unreleased) and i'm _fairly sure_ that it's related to the DSL pkg we're using being the 213 version not 221.

I think we should downgrade this for the time being

After talking with Dan, this might be a compelling reason to switch to the per year branch strategy

jazanne reviewed (commented):
I changed our compatibility range to exclude EAP (221 versions) since they were causing a lot of issues. See inline comment

jazanne committed:
update ide verifier for versions we actually verify against
